### ISSUE TITLE
Redirects not working correctly together with middleware

### ISSUE THREAD

**Issue by serGlazkov** [vercel/next.js#56154]
### Link to the code that reproduces this issue

https://github.com/serGlazkov/next-redirect-middleware-bug

### To Reproduce

- Set up redirects in the next.config.js file.
- Confirm that the redirects are working as expected.
- Add a middleware.ts file to the project.
- Click a link that should trigger a redirect.

### Current vs. Expected behavior

**Actual Result**
The user is redirected to the page, but it's rendered without any properties.

**Expected Result**
The user should be redirected to the intended page with all the expected properties. (Should work the same as without middleware)

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64
Binaries:
  Node: 18.14.2
  npm: 9.5.0
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 13.5.3
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Middleware / Edge (API routes, runtime), Routing (next/router, next/navigation, next/link)

**Comment by broksonic21** [comment]
Related: middleware on page router causes GetStaticProps and GetStaticPaths not to be looked at: [https://github.com/vercel/next.js/issues/56222](https://github.com/vercel/next.js/issues/56222)

**Comment by Konul129** [comment]
redirect works when following these rules : https://nextjs.org/docs/pages/api-reference/next-config-js/redirects

**Comment by serGlazkov** [comment]
> redirect works when following these rules : https://nextjs.org/docs/pages/api-reference/next-config-js/redirects

so, what i did wrong in my example?

**Comment by broksonic21** [comment]
Yep, same issue as mine at https://github.com/vercel/next.js/issues/56222 --- with pages router, simply _having_ a middleware.ts - whether it runs, or matches anythings --- causes Next/Link Urls to a) not check GetStaticPaths b) skip GetStaticProps. Removing middleware.ts is only way to resolve this -> this is definitely a bug in NextJS since v13.4.13-canary.0 and continues until today. It doesn't happen in 13.4.12

**Comment by Anishali2** [comment]
@serGlazkov  @broksonic21  brother I'm facing some issue in next js redirects I need your assistance

I want when user enter url it will move to localhost:3000 but i want it to be move localhost:3000/?loading=home
this behaviour is possible but I don't know why it is not working 

```
  {
        source: '/',
        destination: '/?loading=home',
        permanent: true // tried both boolean values but not work
      }
```
  also tried 
```
    {
        source: '/',
        destination: '/',
        has:[ { type:"query", key:"loading",value:"home" } ],
        permanent: true // tried both boolean values but not work
      }

 ```
again not working
 
 If you guide me I'll appreciate 
 Thanks & Waiting for your response

**Comment by xveganxxxedgex** [comment]
I'm still facing this issue on v14.2.26. I can't seem to find any guidance on how to keep `next.config` redirect functionality while using a `middleware` file. In the middleware, the routes are using the correct destination, but for some reason `NextResponse.next` doesn't seem to use them and just keeps the page on the source route.
